### PR TITLE
API v3 

### PR DESCRIPTION
**Pull request by naveed-ahmad**
### Goals for v3
- [x] Rails 5
- [x] Fix schema 
- [x] Fix primary keys 
- [x] Add support for word-to-word translation for multiple language 
- [x] Remove CPK 
- [x] Improved performance
- [x] Document api using Stoplight
- [ ] API throttling and authentication as final step 
- [x] Migrate data to new database, its huge change and I don't want to touch current DB 

## PENDING 
- [ ] search

### CODE REVIEW

**Comment by ahmedre** [comment]
rebuild

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33046

**Comment by ahmedre** [comment]
rebuild

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33047

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33048

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33049

**Comment by ahmedre** [comment]
rebuild

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33050

**Comment by naveed-ahmad** [comment]
rebuild

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33051

**Comment by naveed-ahmad** [comment]
rebuild

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33052

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33053

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33054

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33055

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33056

**Comment by naveed-ahmad** [comment]
rebuild

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33057

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33058

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33059

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33060

**Comment by naveed-ahmad** [comment]
rebuild

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33061

**Comment by naveed-ahmad** [comment]
rebuild

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33062

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33063

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,19 +1,22 @@
...
   end
 
-  rescue_from APIValidation, :with => :throw_the_error
-
-  rescue_from Apipie::Error, with: :throw_apipie_error
+  rescue_from APIValidation, with: :throw_the_error
 
   private
+  def throw_the_error(error)
+    render :json => {:error => error.message}, :status => :not_found
```
Comment (mmahalwy) [inline app/controllers/application_controller.rb]: should use new ruby way of hashes

File: app/controllers/concerns/language_detection.rb [hunk 2]
```diff
@@ -1,3 +1,4 @@
+#TODO: Really what's going on here :) Need to rewrite
```
Comment (mmahalwy) [inline app/controllers/concerns/language_detection.rb]: This detected the language and offset the indices for search - i beleive

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33066

File: app/controllers/concerns/language_detection.rb (hunk 2 above)
Comment (naveed-ahmad) [inline app/controllers/concerns/language_detection.rb]: Yes but code is kind of encrypted :)

**Comment by ahmedre** [comment]
rebuild

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33071

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33072

**Comment by ahmedre** [comment]
rebuild

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33077

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33090

**Comment by ahmedre** [comment]
rebuild

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33093

**Comment by ahmedre** [comment]
rebuild

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33094

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33189

**Comment by ahmedre** [comment]
rebuild

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33193

**Comment by ahmedre** [comment]
rebuild

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33194

**Comment by ahmedre** [comment]
rebuild

**Comment by ahmedre** [comment]
rebuild

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:32768

**Comment by ahmedre** [comment]
rebuild

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:32769

**Comment by ahmedre** [comment]
rebuild

**Comment by ahmedre** [comment]
rebuild

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:32770

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:32772

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:32782

**Pull request merged**